GPflow/param.py

```diff
@@ -519,54 +526,67 @@ class AutoFlow:
-    >>> result = m._session.run(graph,
-                                feed_dict={x_tf:x,
-                                           y_tf:y,
-                                           m._free_vars:m.get_free_state()})
-
+    >>> result = m.session.run(graph, feed_dict={
+                     x_tf:x, y_tf:y, m._free_vars:m.get_free_state()})
     Not only is the syntax cleaner, but multiple calls to the method will
     result in the graph being constructed only once.
 
     """
 
     def __init__(self, *tf_arg_tuples):
-        # NB. TF arg_tuples is a list of tuples, each of which can be used to
-        # construct a tf placeholder.
+        """
+        :param tf_arg_tuples: list of tuples, each of which can be
+                              used to construct a tensorflow placeholder.
+        """
         self.tf_arg_tuples = tf_arg_tuples
 
     def __call__(self, tf_method):
-        @wraps(tf_method)
-        def runnable(instance, *np_args):
-            storage_name = '_' + tf_method.__name__ + '_AF_storage'
+        @functools.wraps(tf_method)
+        def runnable(instance, *np_args, **kwargs):
+            """
+            AutoFlow function wrapper.
+            """
+            if not set(kwargs.keys()).issubset(['session', 'graph']):
+                raise ValueError('Unknown arguments passed.')
+
+            session = kwargs.get('session')
+            graph = kwargs.get('graph')
+
+            storage_name = ''.join(['_', tf_method.__name__, '_AF_storage'])
             if hasattr(instance, storage_name):
-                # the method has been compiled already, get things out of storage
+                # the method has been compiled already,
+                # get things out of storage
                 storage = getattr(instance, storage_name)
             else:
                 # the method needs to be compiled.
                 storage = {}  # an empty dict to keep things in
                 setattr(instance, storage_name, storage)
-                storage['graph'] = tf.Graph()
-                # storage['session'] = tf.Session(graph=storage['graph'])
-                storage['session'] = session.get_session(
-                    graph=storage['graph'],
-                    output_file_name=settings.profiling.output_file_name + "_" + tf_method.__name__,
-                    output_directory=settings.profiling.output_directory,
-                    each_time=settings.profiling.each_time
-                )
-                with storage['graph'].as_default():
-                    storage['tf_args'] = [tf.placeholder(*a) for a in self.tf_arg_tuples]
+                if session is None:
+                    filename = ''.join([settings.profiling.output_file_name,
+                                        '_', tf_method.__name__])
+                    session = session_mngr.get_session(
+                        graph=graph, output_file_name=filename)
+                graph = session.graph
```

Comment:
AutoFlow behaviour replicates model behaviour regarding user defined session and graphs. https://github.com/GPflow/GPflow/pull/458/files#diff-43f82af73d60224cbbb560f1ea85cc1dR128

`session` value is always considered first, if it is non None, then `session` will be used. Otherwise autoflow will create new custom session using `graph` argument. If `graph` in turn is None, then session will pick up default tensorflow graph.